### DISCUSSION TITLE
Correct behavior of CLA

### DISCUSSION THREAD

**Discussion by ebullient** [commonhaus/foundation#86] [Consensus Building]
### Proposal

We have a draft of a CLA in place now.

I did just [push updated text without a PR](https://github.com/commonhaus/foundation-draft/commit/88c5c58f674fdaa2591d129da387ff16fe07f380) to test a build mechanic: 

- We can keep a file in the top-level administrivia repo for projects that want to use a CLA:
    https://github.com/commonhaus/foundation-draft/blob/main/.github/project-cla.yaml

    This file lists the project (name and entity, not sure why OpenJSF has those as distinct things, I need to look at more examples and ask the lawyer.. maybe one of you know. For POC, I followed what they did) and an accompanying gist (owned by the commonhaus bot).

- I updated the GH Action here:
    https://github.com/commonhaus/foundation-draft/blob/main/.github/workflows/cla.yaml
    
    It iterates over all of the projects in that YAML file and creates a temporary file with the  Project Name/Entity and the text of the (common) CLA. It then updates the specific gist with that content.

This means we can use common CLA text (reviewed approved collectively as policy) and have it applied per project (preferred by some employers: permission to work on one project does not imply permission to work on another).

### Voting group

@commonhaus/founders

### Consensus mechanism

Please use emoji reactions to indicate your support:
- **👍 (:+1:)** if it looks good to you
- **👀 (:eyes:)** if you're "ok" with it (it may not be your favorite)
- **👎 (:-1:)** if you think this needs discussion first

Ask questions or raise concerns in comments.

**Comment by maxandersen** [comment]
I probabby missed it but why is it a DCO is not enough?

**Reply by ebullient** [reply]
Projects are not required to use this, but:
- some employers require one (and require one per project they have agreed to let employee work on), and 
- some projects want one (as it is a clear agreement of who owns IP).

I believe (also) that you technically need all commits to be signed for the DCO to apply (certification/attestation per commit). CLA would be signed once (by the contributor), and would apply for all commits by the individual contributor.

The generated gists should work with https://cla-assistant.io/

I question the click-bait of "better", but.. 
https://opensource.com/article/18/3/cla-vs-dco-whats-difference

**Reply by ebullient** [reply]
I'm wrong about the must sign the commits part. ;) NAL

The rest stands.

**Reply by aalmiray** [reply]
As far as I understand DCO applies to every single commit as you mentioned that is, each individual commit must be signed.
CLA applies as a blanket to every contribution made after signing the CLA, thus it only needs to be signed once.

**Reply by maxandersen** [reply]
I do not think that is correct. 

DCO or CLA can both just be signed once. The difference is mainly that DCO only relates to identify the developer origin and CLA in addition typically outline some context on handing over IP to a company or some other entity. 

Signing of commits is just a way to link your commits to the right DCO/CLA. 

Thats why I assume it would be a DCO rather than CLA commonhaus would use. 

Ref: https://en.m.wikipedia.org/wiki/Developer_Certificate_of_Origin
And
https://en.m.wikipedia.org/wiki/Contributor_License_Agreement

**Reply by aalmiray** [reply]
I see now. I'm mostly familiar with DCOs checked via GitHub apps on every commit and thus assumed that was the case for all.

**Reply by ebullient** [reply]
Commonhaus is providing common language for a CLA, but would not require projects to use it. (Other foundations do this as well).

**Reply by maxandersen** [reply]
Yes, so that's what I don't get. Why a cla with full IP text and not just DCO ? 

The "some companies require it" I don't grok why companies prefer something more legally complicated. Got some context for that ?

**Reply by ebullient** [reply]
@aalmiray ? You use a CLA with JReleaser..

**Reply by aalmiray** [reply]
Yes, yes I do. I went down the path of CLA because of 2 reasons:

1. I thought DCO had to be validated on every commit (I was wrong).
2. CLA states the flow of IP rights whereas DCO doesn't. 

I knew at some point JReleaser might join a foundation and/or require a license change thus CLA was the better choice.

**Reply by ebullient** [reply]
Ok. I think it is clear [here](https://github.com/commonhaus/foundation-draft/blob/main/policies/ip-policy.md#contributions) that I am not imposing a CLA requirement on projects. 

I don't want to argue over why a project would choose to use a CLA, perhaps open another discussion for that? (General)

Mechanically, I think what I have works as a first pass / idea. We'll figure out the rest as we go through the bootstrapping phase. 

One thing I don't know, Andres, is if it is possible to change the linked gist with cla-assistant without losing your history. I also don't know (off the top of my head) if it is possible to allow someone else to edit a gist. What I created above works for gists owned by the commonhaus bot user..

**Reply by aalmiray** [reply]
As far as I recall gists are backed by git repos and thus history should be available one way or another. 

Now, who is allowed to edit a gist, have not look too much into it.

**Reply by ebullient** [reply]
we'll figure it out. ;)

**Comment by ebullient** [comment]
Closing for now, as we have something in place. Will revisit when we know more.
